Documentation for stacks, stack components and flavors (#607)
zenml-io/zenml: docs/book/stack-components/secrets-manager.md

---

> description: Using secrets across your ZenML pipeline

Reviewer: ```suggestion
description: Using secrets across your ZenML pipeline.
```

---

Most projects involving either cloud infrastructure or of a certain complexity

> will involve secrets of some kind. ZenML provides you a Secret Manager as 

Reviewer: ```suggestion
will involve secrets of some kind. ZenML provides you a Secrets Manager as 
```

stack component to help manage and use these secrets for your pipeline runs.

{% hint style="warning" %}
Before reading this chapter, make sure that you are familiar with the 
concept of [stacks, stack components and their flavors](./introduction.md).  
{% endhint %}

## Base Abstraction

> The secret manager acts as the one-stop shop for all of the secrets your 

Reviewer: ```suggestion
The secrets manager acts as the one-stop shop for all of the secrets to which your 
```

> pipeline or stack components might need access to. 

Reviewer: ```suggestion
pipeline or stack components might need access. 
```

1. As it is the base class for a specific type of StackComponent,
    it inherits from the StackComponent class. This sets the `TYPE`
    variable to the Secrets manager . The `FLAVOR` class variable needs to be 
    set in the specific subclass.

> 2. The BaseSecretsManager implements the interface for a set of crud

Reviewer: ```suggestion
2. The BaseSecretsManager implements the interface for a set of CRUD
```

from typing import ClassVar, List

from zenml.enums import StackComponentType
from zenml.secret.base_secret import BaseSecretSchema
from zenml.stack import StackComponent

class BaseSecretsManager(StackComponent, ABC):
    """Base class for all ZenML secrets managers."""

    # Class configuration
    TYPE: ClassVar[StackComponentType] = StackComponentType.SECRETS_MANAGER
    FLAVOR: ClassVar[str]

    @abstractmethod
    def register_secret(self, secret: BaseSecretSchema) -> None:
        """Registers a new secret."""

    @abstractmethod
    def get_secret(self, secret_name: str) -> BaseSecretSchema:
        """Gets the value of a secret."""

    @abstractmethod
    def get_all_secret_keys(self) -> List[str]:
        """Get all secret keys."""

    @abstractmethod
    def update_secret(self, secret: BaseSecretSchema) -> None:
        """Update an existing secret."""

    @abstractmethod
    def delete_secret(self, secret_name: str) -> None:
        """Delete an existing secret."""

    @abstractmethod
    def delete_all_secrets(self, force: bool = False) -> None:
        """Delete all existing secrets."""
```

## List of available secrets managers

> Out of the box, ZenML comes with a `LocalSecretsManager` implementation, which 

Reviewer: ```suggestion
Out-of-the-box, ZenML comes with a `LocalSecretsManager` implementation, which 
```

Since you do it this way elsewhere :)

| GCPSecretsManager                                                                                                                                         | gcp_secrets_manager  | gcp_secrets_manager |

If you would like to see the available flavors for secret managers, you can 
use the command:

```shell
zenml secrets-manager flavor list
```

## Build your own custom secrets manager

If you want to create your own custom flavor for a secrets manager, you can 
follow the following steps:

1. Create a class which inherits from the `BaseSecretsManager`.
2. Define the `FLAVOR` class variable.
3. Implement the `abstactmethod`s based on the API of your desired secrets 
manager

Once you are done with the implementation, you can register it through the CLI 
as:

```shell
zenml secrets-manager flavor register <THE-SOURCE-PATH-OF-YOUR-SECRETS-MANAGER>
```

# Some additional implementation details

Different providers in the space of secrets manager have different definitions 
of what constitutes a secret. While some providers consider a single key value 
pair a secret: `'secret_name': 'secret_value'`, other providers have a slightly 

> different definition. For them, a secret is a collection of key value pairs:

Reviewer: ```suggestion
different definition. For them, a secret is a collection of key-value pairs:
```

`{'some_username': 'user_name_1', 'some_pwd': '1234'}`.

ZenML falls into the second category. The implementation of the different 
methods should reflect this convention. In case the specific implementation 
interfaces with a secrets manager that uses the other definition of a secret, 
working with tags can be helpful. See the `GCPSecretsManager` for inspiration.

## SecretSchemas

> One way the ZenML expands on the notion of secrets as dictionaries, is the 

Reviewer: ```suggestion
One way the ZenML expands on the notion of secrets as dictionaries is the 
```

secret schema. A secret schema allows the user to create and use a specific 

> template. A schema could for example require the combination of a username,

Reviewer: ```suggestion
template. A schema could, for example, require the combination of a username,
```
